### PR TITLE
Dump device code

### PR DESCRIPTION
**Pull request by srividya-sundaram**

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (srividya-sundaram) [inline clang/lib/Driver/ToolChains/Clang.cpp:1]: test

**Review by srividya-sundaram — COMMENTED** [review]
test

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -9933,6 +9933,21 @@ void SPIRVTranslator::ConstructJob(Compilation &C, const JobAction &JA,
         TCArgs.MakeArgString("--out-file-list=" + OutputFileName));
     ForeachArgs.push_back(
         TCArgs.MakeArgString("--out-replace=" + OutputFileName));
+    // If fsycl-dump-device-code is passed, put the output files from llvm-spirv
```
Comment (srividya-sundaram) [inline clang/lib/Driver/ToolChains/Clang.cpp:9936]: `SPIRVTranslator::ConstructJob` seems to construct the `llvm-foreach` and `llvm-spirv` command, that generates the `.spv` files.
`llvm-foreach` has a `--out-dir` option using which users can specify the directory for output files; If unspecified,           system temporary directory is assumed.

I tried adding this option `--out-dir=/test`, and the .spv files got put in the test directory.

In the code below, I'm trying to get the path value provided in fsycl-dump-device-code and set is as the `--out-dir` value 
Is this approach ok?

File: clang/lib/Driver/Compilation.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ Compilation::Compilation(const Driver &D, const ToolChain &_DefaultToolChain,
 Compilation::~Compilation() {
   // Remove temporary files. This must be done before arguments are freed, as
   // the file names might be derived from the input arguments.
-  if (!TheDriver.isSaveTempsEnabled() && !ForceKeepTempFiles)
+  if ((!TheDriver.isSaveTempsEnabled() || !TheDriver.isDumpDeviceCodeEnabled()) && !ForceKeepTempFiles)
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Compilation.cpp:50]: I tried to mimic the logic from save-temps, since we want the .`spv` files to be saved and not deleted after the compilation.
But I don't see the .spv saved with this logic. 
Is this approach ok?

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -6162,6 +6175,10 @@ class OffloadingActionBuilder final {
       bool HasSYCLTargetsOption =
           SYCLAddTargets || SYCLTargets || SYCLLinkTargets;
 
+      // Make -fintelfpga flag imply -fsycl.
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:6178]: // Unrelated change. Please ignore.

File: clang/lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -6162,6 +6175,10 @@ class OffloadingActionBuilder final {
...
+      if (SYCLfpga && !HasValidSYCLRuntime)
+        HasValidSYCLRuntime = true;
+
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:6181]: // Unrelated change. Please ignore.

File: clang/lib/Driver/Driver.cpp [hunk 5]
```diff
@@ -6154,6 +6166,7 @@ class OffloadingActionBuilder final {
       Arg *SYCLTargets =
               C.getInputArgs().getLastArg(options::OPT_fsycl_targets_EQ);
       Arg *SYCLAddTargets = Args.getLastArg(options::OPT_fsycl_add_targets_EQ);
+      Arg *SYCLfpga = C.getInputArgs().getLastArg(options::OPT_fintelfpga);
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:6169]: // Unrelated change. Please ignore.

File: clang/lib/Driver/Driver.cpp [hunk 6]
```diff
@@ -1073,6 +1073,12 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
                                false) ||
       C.getInputArgs().hasArg(options::OPT_fsycl_device_only);
 
+  Arg *SYCLfpga = C.getInputArgs().getLastArg(options::OPT_fintelfpga);
```
Comment (srividya-sundaram) [inline clang/lib/Driver/Driver.cpp:1076]: // Unrelated change. Please ignore.

File: clang/lib/Driver/Compilation.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Compilation.cpp:50]: This approach seems aggressive to me.  Dumping device code isn't equivalent to -save-temps.  We should keep the behavior separate, what do you think?

Comment (srividya-sundaram) [reply clang/lib/Driver/Compilation.cpp:50]: >> Dumping device code isn't equivalent to -save-temps.
It is not. I was just trying to see if I can save the .spv files that I was able to redirect using the --out-dir option in llvm-foreach.
If dump device code option is passed, the .spv files are put in the given path. With this code, I was trying to save them, otherwise, they get deleted when compilation exits.

Comment (srividya-sundaram) [reply clang/lib/Driver/Compilation.cpp:50]: Also, save-temps, saves all the intermediate files.
But with dump device code, I'm only trying to save the .spv files.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 1 above)
Comment (srividya-sundaram) [reply clang/lib/Driver/ToolChains/Clang.cpp:9936]: Can you comment on this?

File: clang/lib/Driver/Compilation.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Compilation.cpp:50]: Okay - I think your logic is not quite right, try this
```suggestion
  if (!(TheDriver.isSaveTempsEnabled() || TheDriver.isDumpDeviceCodeEnabled()) && !ForceKeepTempFiles)
```

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 1 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:9936]: Sure, this is fine for redirecting the produced output from the translator into a specific directory.

File: clang/lib/Driver/Compilation.cpp (hunk 2 above)
Comment (srividya-sundaram) [reply clang/lib/Driver/Compilation.cpp:50]: Thank you! The .spv files now get saved in the given directory.
